gitlab/resource_gitlab_project_protected_environment_test.go

```diff
@@ -0,0 +1,239 @@
+
+func testAccCheckGitlabProjectProtectedEnvironmentAttributes(pt *gitlab.ProtectedEnvironment, want *testAccGitlabProjectProtectedEnvironmentExpectedAttributes) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if pt.Name != want.Name {
+			return fmt.Errorf("got name %q; want %q", pt.Name, want.Name)
+		}
+
+		if pt.DeployAccessLevels[0].AccessLevel != accessLevelID[want.CreateAccessLevel] {
+			return fmt.Errorf("got Create access levels %q; want %q", pt.DeployAccessLevels[0].AccessLevel, accessLevelID[want.CreateAccessLevel])
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckGitlabProjectProtectedEnvironmentDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*gitlab.Client)
+	var project string
+	var environment string
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type == "gitlab_project" {
+			project = rs.Primary.ID
+		} else if rs.Type == "gitlab_project_protected_environment" {
+			environment = rs.Primary.ID
+		}
+	}
+
+	pt, response, err := conn.ProtectedEnvironments.GetProtectedEnvironment(project, environment)
+	if err == nil {
+		if pt != nil {
+			return fmt.Errorf("project Protected Environment %s still exists", environment)
+		}
+	}
+	if response.StatusCode != 404 {
+		return err
+	}
+	return nil
+}
+
+func testAccGitlabProjectProtectedEnvironmentConfig(rInt int, postfix string) string {
+	return fmt.Sprintf(`
+resource "gitlab_project" "foo" {
+  name        = "foo-%[1]d"
+  description = "Terraform acceptance tests"
+
+  # So that acceptance tests can be run in a gitlab organization
+  # with no billing
+  visibility_level       = "public"
+  shared_runners_enabled = true
+}
+
+resource "gitlab_project_environment" "env" {
+  project = gitlab_project.foo.id
+  name    = "ProjectProtectedEnvironment-%[1]d-this-suffix-matches-wildcard"
+}
+
+resource "gitlab_project_protected_environment" "ProjectProtectedEnvironment" {
+  depends_on  = [gitlab_project_environment.env]
+  project     = gitlab_project.foo.id
+  environment = "ProjectProtectedEnvironment-%[1]d%[2]s"
+  deploy_access_levels {
+	access_level = "developer"
+  }
+}
+	`, rInt, postfix)
+}
+
+func testAccGitlabProjectProtectedEnvironmentUpdateConfig(rInt int, postfix string) string {
+	return fmt.Sprintf(`
+resource "gitlab_project" "foo" {
+  name        = "foo-%[1]d"
+  description = "Terraform acceptance tests"
+
+  # So that acceptance tests can be run in a gitlab organization
+  # with no billing
+  visibility_level       = "public"
+  shared_runners_enabled = true
+}
+
+resource "gitlab_project_environment" "env" {
+  project = gitlab_project.foo.id
+  name    = "ProjectProtectedEnvironment-%[1]d-this-suffix-matches-wildcard"
+}
+
+resource "gitlab_project_protected_environment" "ProjectProtectedEnvironment" {
+  depends_on  = [gitlab_project_environment.env]
+  project     = gitlab_project.foo.id
+  environment = "ProjectProtectedEnvironment-%[1]d%[2]s"
+  deploy_access_levels {
+    access_level = "maintainer"
+  }
```

Comment:
Can you add test steps for:
1. Having more than once `deploy_access_levels` blocks
2. Using `group_id` and `user_id`.

Reply:
Done in 4d9f680437b362b3a2e3071a9e0a1a362fc58672